pandas/core/arrays/categorical.py

```diff
@@ -852,9 +852,9 @@ def set_categories(self, new_categories, ordered=None, rename=False,
             if (cat.dtype.categories is not None and
                     len(new_dtype.categories) < len(cat.dtype.categories)):
                 # remove all _codes which are larger and set to -1/NaN
-                self._codes[self._codes >= len(new_dtype.categories)] = -1
+                cat._codes[cat._codes >= len(new_dtype.categories)] = -1
         else:
-            codes = _recode_for_categories(self.codes, self.categories,
+            codes = _recode_for_categories(cat.codes, cat.categories,
```

Comment:
This change shouldn't change any existing behavior, as `self` and `cat` should be identical within this branch of the `if`/`else`.  More so a defensive future-proofing change in case future modifications cause a divergence between `self` and `cat` prior to this.  All operations at this point should be in relation to `cat` anyways, as that's the object we'll be returning.